text/0111-feature-flags.md

```diff
@@ -0,0 +1,167 @@
+
+Examples of changes that would not require a feature flag:
+
+* Rephrasing, reorganizing or refactoring
+* Addition or removal of warnings
+* Additions that strictly improve objective, numerical quality criteria (speedup, better browser support)
+* Additions only likely to be noticed by other platforms implementing LWC, invisible to component authors.
+* Polyfilling new language and browser features after they get standardized.
+
+## Detailed Design
+
+### Using features in Code
+
+In your code, you must import bindings from the config package to branch your code. We have two options:
+
+#### If/Then/Else Branching
+
+```js
+import { ENABLE_FOO } from "@lwc/config";
+if (ENABLE_FOO) {
+    runExtra();
+}
+```
+
+which gets compiled to one of the following options:
+
+```js
+import { ENABLE_FOO } from "@lwc/config";
+// foo feature is configured as `true`
+{ // this block is needed in case the then-block declares some binding
+    runExtra();
+}
+```
+
+```js
+import { ENABLE_FOO } from "@lwc/config";
+// foo feature is configured as `false`
+```
+
+```js
+import { ENABLE_FOO } from "@lwc/config";
+// foo feature is configured as `null`
+if (LWC_config.feature.ENABLE_FOO === true) {
+    runExtra();
+}
+```
+
+#### Arrow Function Branching
+
+```js
+import { feature } from "@lwc/config";
+feature('foo', _ => {
+    // do something...
+});
+```
+
+The output is the exact same as the previous option, but overall, this feature has one main benefit, it is a lot easier to identify, since the syntax is always the same.
+
+### Enabling At Build Time
+
+TBD
+
+### Enabling At Runtime
+
+This could be achieve via the global LWC configuration, e.g.:
+
+```js
+// client side configuration
+LWC_config = {
+    features: {
+        foo: true,
+        bar: false,
+    },
+};
+```
+
+All runtime-evaluated features are disabled by default. Therefore, features can only be enabled when the value is `true`.
+
+It is important to notice that this configuration is a global variable, and must be evaluated before LWC runs.
+
+It is also important to notice that only features that are configured as `null` on the LWC multi-package repository can be enabled or disabled on the client side. If the feature is marked as `true`, it will be always present independently of the `LWC_config` value, while those marked as `false` are just not present at runtime, which means attempting to configure those via `LWC_config` will do nothing.
+
+#### Testing
+
+During the test phase, all features will be compiled out as runtime optional (equivalent to setting them to `null` value). This guarantees that we can have different tests for the different branching logic. With that in mind, you can use a test helper function to enable certain features per test, e.g.:
+
+```js
+import { enableFeature } from '@salesforce/lwc-test-support'; // bikeshed required here
+
+enableFeature('foo'); // enabling feature foo for this test (we don't need a way to disable them)
```

Comment:
I suspect this is only useful if we never use a feature flag during the initialization of the engine, e.g.:

```js
const something = FeatureFoo ? () => 1 : () => 2;
```

but it could be used as:

```js
const something = () => FeatureFoo ? 1 : 2;
```

that's a small detail, but probably should be mentioned, otherwise you have no way to test that feature flag.